File: Python/9663 - N-Queen.py
def bt(y):  # back tracking
    global n, ans, col, dia1, dia2

    # base case: end of the row
    if y == n:
        ans += 1
        return

    for x in range(n):
        # no overlap in either direction
        # deploy queen
        if col[x] is False and dia1[x+y] is False and dia2[x-y+n-1] is False:
            col[x] = True
            dia1[x+y] = True
            dia2[x-y+n-1] = True

            bt(y+1)

            # backtracking
            col[x] = False
            dia1[x + y] = False
            dia2[x - y + n - 1] = False


n = int(input())
ans = 0
# determine if it is in the same column
col = [False for _ in range(n)]
# determine if it is in the same diagonal whose slope is positive
dia1 = [False for _ in range(2*n)]
# determine if it is in the same diagonal whose slope is positive
dia2 = [False for _ in range(2*n)]

bt(0)
print(ans)

File: C++/9663 - N Queen.cpp
#include<bits/stdc++.h>
using namespace std;

const int MAX_N = 4 * 1e4;

bool col[MAX_N], dia1[MAX_N], dia2[MAX_N];

int ans, n;

void backTracking(int y){
    if(y == n){
        ans += 1;
        return;
    }
    
    for(int x = 0; x < n; ++x){
        bool possibleToMove = !col[x] && !dia1[x+y] && !dia2[x-y+n-1];
        if(possibleToMove){
            col[x]  = true;
            dia1[x + y] = true;
            dia2[x - y + n - 1] = true;
            
            backTracking(y + 1);
            
            col[x]  = false;
            dia1[x + y] = false;
            dia2[x - y + n - 1] = false;
        }
    }
    
}
int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(nullptr);
    
   
    cin >> n;
    backTracking(0);
    cout << ans << "\n";
    
    return 0;
}
